### DISCUSSION TITLE
Remote INDI with Docker webserver - chicken and egg?

### DISCUSSION THREAD

**Discussion by gingerbreadassassin** [aaronwmorris/indi-allsky#1749] [Q&A] [answered]
Greetings!

I have indi-allsky running on a RPi 4 with a ZWO ASI676MC. I'm trying to get the image processing settings dialed in, but the Pi4 makes this process quite slow. So, I thought I might run the web interface on another machine (running Almalinux) and connect to indiserver on the Pi. I was able to get the gunicorn and mariadb (had to edit `Dockerfile.mariadb` so that it pulls from `mariadb:10`; 11 refused to start, saying something about not being able to initialize system tables) containers to start. I'm using traefik, so omitted the apache/nginx webserver container. 

My issue is that I'm unable to save the camera settings in the dockerized instance:
![image](https://github.com/user-attachments/assets/3e1bb3be-7c40-46a1-99b6-30d322223181)
![image](https://github.com/user-attachments/assets/bf866c54-cba9-4e83-bf82-6895a5989aee)

No fields are highlighted, but this stacktrace shows up in the gunicorn logs:
<details>
  <summary>stacktrace</summary>

```
gunicorn.indi.allsky-1  | --- Logging error ---
gunicorn.indi.allsky-1  | Traceback (most recent call last):
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/logging/handlers.py", line 1001, in emit
gunicorn.indi.allsky-1  |     self.socket.send(msg)
gunicorn.indi.allsky-1  | OSError: [Errno 9] Bad file descriptor
gunicorn.indi.allsky-1  | 
gunicorn.indi.allsky-1  | During handling of the above exception, another exception occurred:
gunicorn.indi.allsky-1  | 
gunicorn.indi.allsky-1  | Traceback (most recent call last):
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/logging/handlers.py", line 869, in _connect_unixsocket
gunicorn.indi.allsky-1  |     self.socket.connect(address)
gunicorn.indi.allsky-1  | FileNotFoundError: [Errno 2] No such file or directory
gunicorn.indi.allsky-1  | 
gunicorn.indi.allsky-1  | During handling of the above exception, another exception occurred:
gunicorn.indi.allsky-1  | 
gunicorn.indi.allsky-1  | Traceback (most recent call last):
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/logging/handlers.py", line 1004, in emit
gunicorn.indi.allsky-1  |     self._connect_unixsocket(self.address)
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/logging/handlers.py", line 880, in _connect_unixsocket
gunicorn.indi.allsky-1  |     self.socket.connect(address)
gunicorn.indi.allsky-1  | FileNotFoundError: [Errno 2] No such file or directory
gunicorn.indi.allsky-1  | Call stack:
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/threading.py", line 1002, in _bootstrap
gunicorn.indi.allsky-1  |     self._bootstrap_inner()
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
gunicorn.indi.allsky-1  |     self.run()
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/threading.py", line 982, in run
gunicorn.indi.allsky-1  |     self._target(*self._args, **self._kwargs)
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 83, in _worker
gunicorn.indi.allsky-1  |     work_item.run()
gunicorn.indi.allsky-1  |   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
gunicorn.indi.allsky-1  |     result = self.fn(*self.args, **self.kwargs)
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/gunicorn/workers/gthread.py", line 281, in handle
gunicorn.indi.allsky-1  |     keepalive = self.handle_request(req, conn)
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/gunicorn/workers/gthread.py", line 333, in handle_request
gunicorn.indi.allsky-1  |     respiter = self.wsgi(environ, resp.start_response)
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
gunicorn.indi.allsky-1  |     return self.wsgi_app(environ, start_response)
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
gunicorn.indi.allsky-1  |     response = self.full_dispatch_request()
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
gunicorn.indi.allsky-1  |     rv = self.dispatch_request()
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
gunicorn.indi.allsky-1  |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
gunicorn.indi.allsky-1  |   File "/home/allsky/indi-allsky/indi_allsky/flask/misc.py", line 22, in decorated_view
gunicorn.indi.allsky-1  |     return app.ensure_sync(func)(*args, **kwargs)
gunicorn.indi.allsky-1  |   File "/home/allsky/venv/lib/python3.11/site-packages/flask/views.py", line 110, in view
gunicorn.indi.allsky-1  |     return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
gunicorn.indi.allsky-1  |   File "/home/allsky/indi-allsky/indi_allsky/flask/views.py", line 133, in dispatch_request
gunicorn.indi.allsky-1  |     self.cameraSetup(camera_id=camera_id)
gunicorn.indi.allsky-1  |   File "/home/allsky/indi-allsky/indi_allsky/flask/base_views.py", line 66, in cameraSetup
gunicorn.indi.allsky-1  |     self.camera = self.getCameraById(camera_id)
gunicorn.indi.allsky-1  |   File "/home/allsky/indi-allsky/indi_allsky/flask/base_views.py", line 110, in getCameraById
gunicorn.indi.allsky-1  |     app.logger.warning('No cameras are defined')
gunicorn.indi.allsky-1  | Message: 'No cameras are defined'
gunicorn.indi.allsky-1  | Arguments: ()
```
</details>

As evidenced in the above screenshot, the drop-down selector for camera in the top left is indeed blank. If that needs to be populated before being able to save camera settings, how does one actually connect to an external indiserver instance in order to populate it? Is my dilemma a result of a race condition in form submission/input validation, or did I overlook a configurable somewhere?

**Comment by gingerbreadassassin** [comment]
Made some progress:

1. Exec into the gunicorn container:
`$ docker exec -it docker-gunicorn.indi.allsky-1 /bin/bash`

2. Activate the virtualenv:
`$ source venv/bin/activate`

3. Enter the indi-allsky directory:
`$ cd indi-allsky/`

4. Install an editor (I chose vim, others might prefer nano):
`$ sudo apt update && apt install -y vim`

5. Edit the config with config.py:
`$ python config.py edit`

6. Exit from the container:
`$ exit`

7. Reload the web interface

Now the correct IP address shows up in the config page, but I still see no cameras available. I also don't see anything connecting to indiserver on the pi4:
```
pi@indi-allsky:~/indi-allsky $ sudo ss -atunp | grep indi
tcp   LISTEN    0      5                          0.0.0.0:7624                 0.0.0.0:*     users:(("indiserver",pid=1246,fd=5))                                                   
tcp   ESTAB     3263   0                            [::1]:37534                  [::1]:2947  users:(("indi_gpsd",pid=1250,fd=3))    
```

**Comment by gingerbreadassassin** [comment, marked as answer]
Got it. Had to start the `capture` container as well. Here's what my compose file looks like:

<details>
  <summary>docker-compose.yaml</summary>

```
services:
  indi.base:
    build:
      context: ..
      dockerfile: docker/Dockerfile.indi_base_debian12
      args:
        - TZ=${TZ}
        - INDI_VERSION=${INDIALLSKY_INDI_VERSION}
        - LIBCAMERA_TAG=${INDIALLSKY_LIBCAMERA_TAG}
        - RPICAM_APPS_TAG=${INDIALLSKY_RPICAM_APPS_TAG}
    env_file: .env


  mariadb.indi.allsky:
    build:
      context: ..
      dockerfile: docker/Dockerfile.mariadb
    env_file: .env
    volumes:
      - database_indi_allsky:/var/lib/mysql
    networks:
      - private


  gunicorn.indi.allsky:
    build:
      context: ..
      dockerfile: docker/Dockerfile.gunicorn
    env_file: .env
    depends_on:
      - mariadb.indi.allsky
    volumes:
      - images_indi_allsky:/var/www/html/allsky
      - migrations_indi_allsky:/var/lib/indi-allsky
    networks:
      - private
      - webgateway
    labels:
      - traefik.enable=true
      - traefik.docker.network=webgateway
      - traefik.http.routers.indi-allsky.rule=Host(`indi-allsky.mydomain.tld`)
      - traefik.http.routers.indi-allsky.entrypoints=websecure
      - traefik.http.routers.indi-allsky.tls=true
      - traefik.http.routers.indi-allsky.tls.certresolver=le
      - traefik.http.routers.indi-allsky.service=indi-allsky
      - traefik.http.services.indi-allsky.loadbalancer.server.port=8000


  capture.indi.allsky:
    build:
      context: ..
      dockerfile: docker/Dockerfile.capture
    env_file: .env
    depends_on:
      - indi.base
      - mariadb.indi.allsky
        #- indiserver.indi.allsky
    networks:
      - private
    volumes:
      - images_indi_allsky:/var/www/html/allsky

volumes:
  database_indi_allsky:
  migrations_indi_allsky:
  images_indi_allsky:
  mosquitto_data_indi_allsky:

networks:
  private:
  webgateway:
    external: true
```
</details>

Notably missing are indiserver, webserver, and mosquitto services/containers. Also, I've left the commented-out `depends_on: indiserver.indi.allsky` for the capture service to intentionally highlight it.

And my new start command:

`docker compose up -d`

**Comment by aaronwmorris** [comment]
The capture container is where the camera would get defined when it initializes the camera for the first time.

I am not sure why you could not save the config.  I am assuming it was a setting that did not pass validation.  The error would have caused the setting to highlight red, but you might have had to click through the different tabs to find it.  It was not the indiserver hostname since that was not highlighted.

I tested a new build this morning and the database container started fine with MariaDB 11.  I am not sure what happened there.

**Reply by gingerbreadassassin** [reply]
I'm not exactly sure about why either. I will say that I tested by stopping all containers, deleting the volumes, and starting from scratch. With the three containers, mariadb, gunicorn, and capture, I was able to actually save the config, despite no camera being present. So the stacktrace I shared was indeed ultimately not the underlying issue preventing the config from being saved.

As for issues with MariaDB 11, I figured it's likely something to do with my particular setup. Whether it's my installation on Almalinux, using SELinux, the version of docker I'm running or the kernel I have installed, I don't know. I even tried the `lts` tag, and it wouldn't start. But I didn't open an issue about this because my setup is fairly far from typical. I just figured I'd mention it here at least for anyone with similar issues.

Ultimately, this was intended to be a temporary setup to allow me to use some beefier hardware to fine-tune the processing on the 12MP images provided by the ASI676MC. CLAHE, stretching, and SNCR on a Pi 4 takes long enough to be annoying when trying to dial things in. I intend simply to copy over the settings when I'm happy with the results, and let the Pi do all the processing in the future.

**Reply by aaronwmorris** [reply]
Sounds good.  Just an FYI, if you enable saving a few FITS files at night, you can use the FITS Viewer view to select a files for the "Process FITS" view to test out different processing settings dynamically.

**Reply by gingerbreadassassin** [reply]
Literally doing this right now :D

**Comment by spikked27** [comment]
I may be doing this completely wrong but I'm attempting to do this within Unraid using the docker compose plugin which I've had success with in the past, after setting up the docker compose yml and env I get the following error when trying to compose up:

![image](https://github.com/user-attachments/assets/c5655901-4816-4891-9d6a-687a6ff3d9f7)

Is there something obvious I'm missing?

**Reply by gingerbreadassassin** [reply]
I'm not familiar with Unraid; I run the containerized version of indi-allsky on a dedicated docker host, which uses NFS shares from my TrueNAS box, but I'll try to help.

What's happening here seems like you've just copy+pasted the compose file into the Unraid webui. I don't think this will work so simply as the compose file actually relies on other files/directories in the git repository to build the images. Someone might be uploading prebuilt images to dockerhub or some other repository that you might be able to point the compose file to (and subsequently drop the 'build' contexts), but I'm unaware of any such images.

**Reply by spikked27** [reply]
You are correct, I cloned the github into the directory the compose file is point to and got further but ended up getting failed to calculate checksum error:

![image](https://github.com/user-attachments/assets/b858f2ad-678d-4c40-a2c9-6fd181e275e8)

**Reply by gingerbreadassassin** [reply]
Failing to calculate the checksum isn't exactly the problem, it's `"/misc/build_indi.sh": not found` 

It's trying to find [this file](https://github.com/aaronwmorris/indi-allsky/blob/main/misc/build_indi.sh), but is unable to. Likely, this is because wherever you cloned the git repository into is not the correct directory or is otherwise inaccessible by the docker build process (e.g. a permissions issue). 

You might try editing each of the "context" contexts in [the compose file](https://github.com/aaronwmorris/indi-allsky/blob/main/docker/docker-compose.yaml#L5) such that they point, fully-qualified (as in, `/the/whole/directory/path/to/indi-allsky`), to wherever you cloned the repository into. Then retry to build / start the stack.

If that fails, you're honestly likely better off building the images manually via CLI, either on unraid if possible, or on another machine and somehow getting those images to your unraid box, then omitting the build context from the compose file while also specifying the prebuilt images. But that's really out-of-scope for this discussion, and you likely would greatly benefit from learning more about how docker and docker compose work _outside_ of an Unraid environment. Maybe spin up a VM on your unraid and run through some docker/docker compose tutorials on it. 

In short: you're looking for a drop-in solution to run in an obfuscated environment, in which indi-allsky is not intended to run. It doesn't mean it can't be run in Unraid, it's just that you have to do some work outside of the webui and have some knowledge about docker/compose to get it to function.

**Reply by aaronwmorris** [reply]
If you are building the docker images manually, make sure you are starting the process in the root of the indi-allsky git checkout, not within the docker folder.  The path to the build_indi.sh script (and others) is from the root of the checkout.
